decoder/jws.go

```diff
@@ -66,6 +73,13 @@ func (d *jwsDecoder) Decode(ctx context.Context, rawJws string) (*Token, error)
 }
 
 func (d *jwsDecoder) parseAndValidate(rawJws string) (jwt.Token, error) {
+	if d.jwks == nil {
+		jwks, err := TryFetchingJwks(d.jwksURL)
+		if err != nil {
+			return nil, fmt.Errorf("jwks: failed to fetch from url %s: %w", d.jwksURL, err)
+		}
+		d.jwks = jwks
```

Comment:
This will have to be protected by a some synchronization to avoid race conditions.

one option would be an rw mutex:
```go
func (d *jwsDecoder) getJWKs() (*jwk.Set, error) {
  d.mutex.RLock()
  if d.jwks != nil {
    d.mutex.RUnlock()
    return d.jwks, nil
  }
  d.mutex.RUnlock()
  d.mutex.Lock()
  defer d.mutex.Unlock()
  if d.jwks != nil {
    return d.jwks, nil
  }
  jwks, err := TryFetchingJwks(d.jwksURL)
  if err != nil {
    return nil, fmt.Errorf("jwks: failed to fetch from url %s: %w", d.jwksURL, err)
  }
  d.jwks = jwks
  return d.jwks, nil
}
```

Reply:
You are totally right, I forgot that go http server is concurrent not single threaded.

Updated the code and simplified it a little more. I don't think we need a separate RLock block, because under normal circumstances, when the JWK set is fetched it shouldn't continue to the write phase. In fact, if it continues to fetching block the reading section becomes redundant.

In addition, I added one test case using googleapis certs to test valid JWKS.